components/calendar/src/chinese_based.rs

```diff
@@ -0,0 +1,346 @@
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! This module contains types and traits for use in the Chinese traditional lunar calendar,
+//! as well as in related and derived calendars such as the Korean and Vietnamese lunar calendars.
+//!
+//! ```rust
+//! use icu::calendar::{chinese::Chinese, Iso, Date};
+//!
+//! let iso_date = Date::try_new_iso_date(2023, 6, 23).unwrap();
+//! let chinese_date = Date::new_from_iso(iso_date, Chinese);
+//! // TODO: When Korean and Vietnamese are added, put example code here as well.
+//!
+//! assert_eq!(chinese_date.year().number, 4660);
+//! assert_eq!(chinese_date.year().related_iso, Some(2023));
+//! assert_eq!(chinese_date.year().cyclic, Some(40));
+//! assert_eq!(chinese_date.month().ordinal, 6);
+//! assert_eq!(chinese_date.day_of_month().0, 6);
+//! ```
+
+use crate::{
+    astronomy::{Astronomical, Location, MEAN_SYNODIC_MONTH, MEAN_TROPICAL_YEAR},
+    calendar_arithmetic::{ArithmeticDate, CalendarArithmetic},
+    helpers::{adjusted_rem_euclid, i64_to_i32, quotient, I32Result},
+    rata_die::RataDie,
+    types::Moment,
+    Calendar, Date,
+};
+
+/// The trait ChineseBased is used by Chinese-based calendars to perform computations shared by such calendar.
+/// To do so, calendars should:
+///
+/// - Implement `fn location` by providing a location at which observations of the moon are recorded, which
+/// may change over time (the zone is important, long, lat, and elevation are not relevant for these calculations)
+/// - Define `const EPOCH` as a `RataDie` marking the start date of the era of the Calendar for internal use,
+/// which may not accurately reflect how years or eras are marked traditionally or seen by end-users
+/// - Implement `fn new_chinese_based_date` by taking a year, month, and day in a Chinese-based calendar and
+/// returning a Date of the relevant Calendar type.
+///
+/// For an example of how to use this trait, see `impl ChineseBased<Chinese>` in [`Chinese`].
+pub(crate) trait ChineseBased<C: CalendarArithmetic> {
+    /// Given a fixed date, return the location used for observations of the new moon in order to
+    /// calculate the beginning of months. For multiple Chinese-based lunar calendars, this has
+    /// changed over the years, and can cause differences in calendar date.
+    fn location(fixed: RataDie) -> Location;
+
+    /// The RataDie of the beginning of the epoch used for internal computation; this may not
+    /// reflect traditional methods of year-tracking or eras, since Chinese-based calendars
+    /// may not track years ordinally in the same way many western calendars do.
+    const EPOCH: RataDie;
+
+    /// Given a year, month, and day, create a Date<C> where C is a Chinese-based calendar.
+    ///
+    /// This function should just call try_new_C_date where C is the name of the calendar.
+    fn new_chinese_based_date(year: i32, month: u8, day: u8) -> Date<C>;
```

Comment:
```suggestion
    fn new_chinese_based_date(year: i32, month: u8, day: u8) -> Date<Self>;
```

Reply:
Done, but this results in an error:

"the size for values of type `Self` cannot be known at compilation time"

I'm not sure how to fix this

Reply:
See latest commit

Reply:
Trying to make that function specifically take a generic type doesn't seem to work either, see the first comment on this PR